Adding the ability to support large file uploads with intermediate saving checkpoints

ivanleomk opened the pull request.

jxnl committed:
improve uploads

jxnl committed:
try hf_transfer

jxnl committed:
update

jxnl committed:
benchmarks

jxnl committed:
split uploads

jxnl committed:
reorg embedding function

jxnl committed:
still does not work

ivanleomk committed:
Added tmp debugging code

ivanleomk committed:
Managed to get HF_Transfer working

jxnl commented on applications/wikipedia/main.py:
```diff
@@ -219,17 +231,26 @@ def embed_dataset(down_scale: float = 0.005, batch_size: int = 512 * 50):
+        os.makedirs(checkpoint_dir,exist_ok=True)
+        pq.write_table(table, path)
+        del dataset
+        for _ in range(3):
```
Do we still need this

ivanleomk replied:
Have removed it in the latest commit. I'm just working with a separate volume now. `del dataset` seems to not work, not too sure why. I tried deleting the variable immediately after generating the batches but it still doesn't release the lock on the file

jxnl commented on applications/wikipedia/upload.py:
```diff
@@ -0,0 +1,63 @@
+    start = time.perf_counter()
+    # dataset.push_to_hub(dataset_name)
+    api.upload_folder(
+        folder_path=f"{cache_dir}/wikipedia",
```
Can you update this to the checkpoints and then try a 25% batch and see how long it takes.

ivanleomk replied:
```
  step b1a6b6e1eff0e9e0fa2a6d579b8f38b93af9bec1e499c724818f11cc2186b2c0 completed (still 1 to go).
data-00124-of-00127.arrow: 512MB [00:01, 512MB/s]                           
data-00125-of-00127.arrow: 512MB [00:10, 51.1MB/s]                          
data-00126-of-00127.arrow: 512MB [00:01, 401MB/s]                           
100%|██████████| 3/3 [00:12<00:00,  4.27s/it]
  step c9c3d34719f6b51dac2a4c29e9b3b20a440bb1f40a7fa669e18185dbb1aacf9e completed (still 0 to go).
All commits have been pushed.
PR is now open for reviews.
PR has been automatically merged (`merge_pr=True` was passed).
Uploaded in 758.460917796s
```

It took about 12.5 minutes to upload the entire dataset. I've found that upload time doesn't scale as much as the absolute size of the dataset though, 10% of dataset took 10 minutes while 25% took 12 minutes. We might be able to stay < 30 mins for uploading the entire dataset of embeddings ( optimistically )

ivanleomk committed:
main.py now embeds and saves to HF

jxnl commented on applications/wikipedia/download.py:
```diff
@@ -51,9 +51,15 @@ def check_dataset_exists():
     print("Dataset loaded.")
 
+    # Now try to save to disk
+    with open(f"{cache_dir}/new_text_file.txt", "w") as file:
```
lets clean this up

jxnl commented on applications/wikipedia/main.py:
```diff
@@ -20,13 +20,18 @@
-dataset_file = "wiki-embeddings.parquet"
+SAVE_TO_DISK = True
+dataset_name = f"567-labs/wikipedia-embedding-{MODEL_SLUG}-debug"
+hf_dataset_name = "567-labs/upload-25"
```
lets clean this up and prepare for a 100%

jxnl commented on applications/wikipedia/main.py:
```diff
@@ -219,17 +232,38 @@ def embed_dataset(down_scale: float = 0.005, batch_size: int = 512 * 50):
+
+        print(f"Pushing to hub {hf_dataset_name}")
+        start = time.perf_counter()
+        api.upload_folder(
```
does this mean we don't need upload,py?

ivanleomk committed:
Fixed up main.py

ivanleomk committed:
Added two additional sections for the dataset processing

ivanleomk committed:
Merge branch 'main' into paraquet-saving

ivanleomk committed:
Fixed a merge conflict and updated embedding article to reflect it

ivanleomk committed:
Refactored some of the code to be cleaner

ivanleomk committed:
Added examples for how to use the Modal volume

ivanleomk commented:
Made the following changes

- deleted `local.py` and `upload.py` since I don't think they're needed now
- refactored out some of the logic into helper functions so that it reads a bit more easily
- Cleaned up the vars in `main.py` so that config and constants variables are all capitalised by default 
- cleaned up `download.py` and removed the useless functions.

Also added an example of how to use the Modal cli `volume` feature with an example for `ls` and `rm` inside the Readme file.
